pallets/parachain-staking/src/lib.rs

```diff
@@ -825,30 +823,61 @@ pub mod pallet {
 		/// - The transaction's complexity is mainly dependent on updating the
 		///   `SelectedCandidates` storage in `select_top_candidates` which in
 		///   return depends on the number of `MaxSelectedCandidates` (N).
-		/// - For each N, we read `CollatorState` from the storage.
+		/// - For each N, we read `CandidatePool` from the storage.
 		/// ---------
-		/// Weight: O(N) where N is `MaxSelectedCandidates` bounded by
-		/// `MaxTopCandidates`
-		/// - Reads: MaxSelectedCandidates, TopCandidates, N * CollatorState
+		/// Weight: O(N + D) where N is `MaxSelectedCandidates` bounded by
+		/// `MaxTopCandidates` and D is the number of delegators of a
+		/// candidate bounded by `MaxDelegatorsPerCollator`.
+		/// - Reads: MaxSelectedCandidates, TopCandidates, N * CandidatePool
 		/// - Writes: MaxSelectedCandidates
 		/// # </weight>
-		#[pallet::weight(<T as pallet::Config>::WeightInfo::set_max_selected_candidates(*new, (*new).saturating_mul(T::MaxDelegatorsPerCollator::get())))]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::set_max_selected_candidates(
+			*new,
+			T::MaxDelegatorsPerCollator::get()
```

Comment:
Q: I can't see where the second argument for weight computation is useful. We compute it down here, but we always take the max and we never iterate over those, only compute the `len()` of the list, so I don't think it's relevant neither for the benchmark nor to keep track of it in any way?

Reply:
I hoped to get rid of the required number of delegator arguments for most benchmarks but unfortunately they are still necessary. However, before we took the sum of the delegators of all the collators we iterate in an extrinsic over as the value which was incorrect. Ideally, we would take something the average number of delegators each collator, which is relevant for the call, has. For worst case handling, we take the maximum number of delegators of a collator here.

I don't see why this should not be relevant? And what else should we do other than taking the `len()` of the list? This directly tells us the number of delegators for a specific collator which is all that we are interested in for this parameter.

Reply:
The other option would be to run the benchmarks always with the maximum number of delegators.

Reply:
What I am saying is that, taking the `set_max_selected_candidates(origin, new)` as example, the only places where we use the information about the number of delegators are the following:

- when iterating over `TopCandidates`, where we take, for each candidate, its `state.delegators.len().saturated_into::<u32>()`;
- in the `reduce`, we then take the max value for that field among all the collators we iterate over

Once that is done, we update `TotalCollatorStake`, where we don't use the information about the number of delegators.
What I am saying is that we are not iterating over the number of delegators we get from the iteration, in this specific extrinsic, so unless the `.len()` function is dependent on the size of the delegators set, I do not see how the max number of delegators we find for a collator affects the runtime execution in any way.
Am I missing something else?

Reply:
As discussed, I suppose the number of delegators factor in because when reading the state of a candidate from `CandidatePool`, the delegations still have to be decoded.